imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -988,3 +1026,270 @@ def create_datasets(
+    """
+    Initialize the L1B histogram rates dataset.
+
+    Parameters
+    ----------
+    l1a_hist : xr.Dataset
+        The L1A histogram rates dataset.
+    attr_mgr_l1b : ImapCdfAttributes
+        Attribute manager used to get the L1B histogram rates dataset attributes.
+    logical_source : str
+        The logical source of the data product that's being created.
+
+    Returns
+    -------
+    l1b_histrates : xr.Dataset
+        The initialized L1B histogram rates dataset.
+    """
+    l1b_histrates = xr.Dataset(
+        attrs=attr_mgr_l1b.get_global_attributes(logical_source),
+    )
+
+    l1b_histrates["epoch"] = xr.DataArray(
+        l1a_hist["epoch"].values,
+        dims=["epoch"],
+        attrs=attr_mgr_l1b.get_variable_attributes("epoch"),
+    )
+    # Copy over fields from L1A DE that will not change in L1B processing
+    l1b_histrates["h_counts"] = xr.DataArray(
+        l1a_hist["hydrogen"].values,
+        dims=["epoch"],
+        # TODO: Add hydrogen to YAML file
+        # attrs=attr_mgr.get_variable_attributes("hydrogen"),
+    )
+    l1b_histrates["o_counts"] = xr.DataArray(
+        l1a_hist["oxygen"].values,
+        dims=["epoch"],
+        # TODO: Add oxygen to YAML file
+        # attrs=attr_mgr.get_variable_attributes("oxygen"),
+    )
+
+    return l1b_histrates
+
+
+def resweep_histogram_data(
+    l1b_histrates: xr.Dataset,
+    anc_dependencies: list,
+) -> xr.Dataset:
+    """
+    Correct energy steps in histogram data based on sweep and LUT tables.
+
+    This function maps the nominal energy step indices (1-7) to actual energy
+    step levels from the LUT table, consolidating counts where multiple indices
+    map to the same energy level.
+
+    Parameters
+    ----------
+    l1b_histrates : xr.Dataset
+        The L1B histogram rates dataset containing h_counts and o_counts.
+    anc_dependencies : list
+        List of ancillary file paths.
+
+    Returns
+    -------
+    l1b_histrates : xr.Dataset
+        Updated dataset with corrected energy step mappings.
+    """
+    sweep_df = lo_ancillary.read_ancillary_file(
+        next(str(s) for s in anc_dependencies if "sweep-table" in str(s))
+    )
+    lut_df = lo_ancillary.read_ancillary_file(
+        next(str(s) for s in anc_dependencies if "esa-mode-lut" in str(s))
+    )
+
+    # Convert sweep table dates to ttj2000ns for direct comparison with epochs
+    sweep_dates = sweep_df["Date"].astype(str)
+
+    # Create mapping from date to LUT table for faster lookup
+    date_to_lut = dict(zip(sweep_dates, sweep_df["LUT_table"], strict=False))
+
+    # Get epoch values
+    epochs = l1b_histrates["epoch"].values
+
+    # Initialize corrected data arrays
+    h_counts_reswept = np.zeros_like(l1b_histrates["h_counts"].values)
+    o_counts_reswept = np.zeros_like(l1b_histrates["o_counts"].values)
+
+    # Process each epoch
+    for epoch_idx, epoch_value in enumerate(epochs):
+        epoch_date = et_to_utc(ttj2000ns_to_et(epoch_value))
```

Comment:
I think these are vectorized functions, so we should calculate the list of dates all at once up front.

```suggestion
    epoch_dates = et_to_utc(ttj2000ns_to_et(epochs))
    for epoch_idx, epoch_date in enumerate(epoch_dates):
        
```